File: cibcc/python/step12/sect2/codegen.py
labelseq = 0
def gen(node):
	global labelseq
	if node.kind == 'NUM':
		print(f"  push {node.val}")
		return
	elif node.kind == 'EXPR_STMT':
		gen(node.lhs)
		print("  add rsp, 8")
		return
	elif node.kind == 'VAR':
		gen_addr(node)
		load()
		return
	elif node.kind == 'ASSIGN':
		gen_addr(node.lhs)
		gen(node.rhs)
		store()
		return
	elif node.kind == 'IF':
		seq = labelseq
		labelseq += 1
		if node.els:
			gen(node.cond)
			print("  pop rax")
			print("  cmp rax, 0")
			print(f"  je .Lelse{seq}")
			gen(node.then)
			print(f"  jmp .Lend{seq}")
			print(f".Lelse{seq}:")
			gen(node.els)
			print(f".Lend{seq}:")
		else:
			gen(node.cond)
			print("  pop rax")
			print("  cmp rax, 0")
			print(f"  je .Lend{seq}")
			gen(node.then)
			print(f".Lend{seq}:")
		return
	elif node.kind == 'WHILE':
		seq = labelseq
		labelseq += 1
		print(f".Lbegin{seq}:")
		gen(node.cond)
		print("  pop rax")
		print("  cmp rax, 0")
		print(f"  je .Lend{seq}")
		gen(node.then)
		print(f"  jmp .Lbegin{seq}")
		print(f".Lend{seq}:")
		return
	elif node.kind == 'RETURN':
		gen(node.lhs)
		print("  pop rax")
		print("  jmp .Lreturn")
		return

	gen(node.lhs)
	gen(node.rhs)

	print("  pop rdi")
	print("  pop rax")

	if node.kind == 'ADD':
		print("  add rax, rdi")
	elif node.kind == 'SUB':
		print("  sub rax, rdi")
	elif node.kind == 'MUL':
		print("  imul rax, rdi")
	elif node.kind == 'DIV':
		print("  cqo")
		print("  idiv rdi")
	elif node.kind == 'EQ':
		print("  cmp rax, rdi")
		print("  sete al")
		print("  movzb rax, al")
	elif node.kind == 'NE':
		print("  cmp rax, rdi")
		print("  setne al")
		print("  movzb rax, al")
	elif node.kind == 'LT':
		print("  cmp rax, rdi")
		print("  setl al")
		print("  movzb rax, al")
	elif node.kind == 'LE':
		print("  cmp rax, rdi")
		print("  setle al")
		print("  movzb rax, al")

	print("  push rax")

def codegen(prog):
	print(".intel_syntax noprefix")
	print(".global main")
	print("main:")

	print("  push rbp")
	print("  mov rbp, rsp")
	print(f"  sub rsp, {prog.stack_size}")

	cur = prog.node
	while cur:
		gen(cur)
		cur = cur.next

	print(".Lreturn:")
	print("  mov rsp, rbp")
	print("  pop rbp")
	print("  ret")

def gen_addr(node):
	if node.kind == 'VAR':
		print(f"  lea rax, [rbp-{node.var.offset}]")
		print("  push rax")
		return
	raise Exception("not an lvalue")  

def load():
	print("  pop rax")
	print("  mov rax, [rax]")
	print("  push rax")

def store():
	print("  pop rdi")
	print("  pop rax")
	print("  mov [rax], rdi")
	print("  push rdi")

File: cibcc/java/step12/sect2/Codegen.java
public class Codegen {

	int labelseq = 0;

	public void codegen(Program prog) {

		System.out.println(".intel_syntax noprefix");
		System.out.println(".global main");
		System.out.println("main:");

		System.out.println("  push rbp");
		System.out.println("  mov rbp, rsp");
		System.out.printf("  sub rsp, %d\n", prog.stack_size);

		for (Node n = prog.node; n != null; n = n.next) {
			gen(n);
		}

		System.out.println(".Lreturn:");
		System.out.println("  mov rsp, rbp");
		System.out.println("  pop rbp");
		System.out.println("  ret");
	}

	private void gen(Node node) {

		if (node.kind == NodeKind.ND_NUM) {
			System.out.printf("  push %d\n", node.val);
			return;
		}
		if (node.kind == NodeKind.ND_EXPR_STMT) {
			gen(node.lhs);
			System.out.println("  add rsp, 8");
			return;
		}
		if (node.kind == NodeKind.ND_VAR) {
			gen_addr(node);
			load();
			return;
		}
		if (node.kind == NodeKind.ND_ASSIGN) {
			gen_addr(node.lhs);
			gen(node.rhs);
			store();
			return;
		}
		if (node.kind == NodeKind.ND_IF) {
			int seq = labelseq;
			labelseq++;
			if (node.els !=null) {
				gen(node.cond);
				System.out.println("  pop rax");
				System.out.println("  cmp rax, 0");
				System.out.printf("  je .Lelse%d\n", seq);
				gen(node.then);
				System.out.printf(".Lelse%d:\n", seq);
				gen(node.els);
			} else {
				gen(node.cond);
				System.out.println("  pop rax");
				System.out.println("  cmp rax, 0");
				System.out.printf("  je .Lend%d\n", seq);
				gen(node.then);
				System.out.printf(".Lend%d:\n", seq);
			}
			return;
		}
		if (node.kind == NodeKind.ND_WHILE) {
			int seq = labelseq++;
			System.out.printf(".Lbegin%d:\n", seq);
			gen(node.cond);
			System.out.println("  pop rax");
			System.out.println("  cmp rax, 0");
			System.out.printf("  je .Lend%d\n", seq);
			gen(node.then);
			System.out.printf("  jmp .Lbegin%d\n", seq);
			System.out.printf(".Lend%d:\n", seq);
			return;
		}
		if (node.kind == NodeKind.ND_RETURN) {
			gen(node.lhs);
			System.out.println("  pop rax");
			System.out.println("  jmp .Lreturn");
			return;
		}

		gen(node.lhs);
		gen(node.rhs);

		System.out.println("  pop rdi");
		System.out.println("  pop rax");

		switch (node.kind) {
			case ND_ADD:
				System.out.println("  add rax, rdi");
				break;
			case ND_SUB:
				System.out.println("  sub rax, rdi");
				break;
			case ND_MUL:
				System.out.println("  imul rax, rdi");
				break;
			case ND_DIV:
				System.out.println("  cqo");
				System.out.println("  idiv rdi");
				break;
			case ND_EQ:
				System.out.println("  cmp rax, rdi");
				System.out.println("  sete al");
				System.out.println("  movzb rax, al");
				break;
			case ND_NE:
				System.out.println("  cmp rax, rdi");
				System.out.println("  setne al");
				System.out.println("  movzb rax, al");
				break;
			case ND_LT:
				System.out.println("  cmp rax, rdi");
				System.out.println("  setl al");
				System.out.println("  movzb rax, al");
				break;
			case ND_LE:
				System.out.println("  cmp rax, rdi");
				System.out.println("  setle al");
				System.out.println("  movzb rax, al");
				break;
		}
		System.out.println("  push rax");
	}

	private void gen_addr(Node node) {
		if (node.kind == NodeKind.ND_VAR) {
			System.out.printf("  lea rax, [rbp-%d]\n", node.va.offset);
			System.out.println("  push rax");
			return;
		}
		System.out.println("not an lvalue");
	}

	private void load() {
		System.out.println("  pop rax");
		System.out.println("  mov rax, [rax]");
		System.out.println("  push rax");
	}

	private void store() {
		System.out.println("  pop rdi");
		System.out.println("  pop rax");
		System.out.println("  mov [rax], rdi");
		System.out.println("  push rdi");
	}
}

	
